internal/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,150 @@
+	*kubeclient.Client,
+	func(context.Context, func(context.Context)),
+	error,
+) {
+	internalClient, err := kubeclient.New(opts...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not create internal client for leader election: %w", err)
+	}
+
+	isLeader := atomic.NewBool(false)
+
+	identity := podInfo.Name
+	leaseName := deployment.Name
+
+	leaderElectionConfig := leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Namespace: podInfo.Namespace,
+				Name:      leaseName,
+			},
+			Client: internalClient.Kubernetes.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: identity,
+			},
+		},
+		ReleaseOnCancel: true, // semantics for correct release handled by controllersWithLeaderElector below
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   15 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(_ context.Context) {
+				plog.Debug("leader gained", "identity", identity)
+				isLeader.Store(true)
+			},
+			OnStoppedLeading: func() {
+				plog.Debug("leader lost", "identity", identity)
+				isLeader.Store(false)
+			},
+			OnNewLeader: func(newLeader string) {
+				if newLeader == identity {
+					return
+				}
+				plog.Debug("new leader elected", "newLeader", newLeader)
+			},
+		},
+		Name: leaseName,
+		// this must be set to nil because we do not want to associate /healthz with a failed
+		// leader election renewal as we do not want to exit the process if the leader changes.
+		WatchDog: nil,
+	}
+
+	// validate our config here before we rely on it being functioning below
+	if _, err := leaderelection.NewLeaderElector(leaderElectionConfig); err != nil {
+		return nil, nil, fmt.Errorf("invalid config - could not create leader elector: %w", err)
+	}
+
+	writeOnlyWhenLeader := kubeclient.MiddlewareFunc(func(_ context.Context, rt kubeclient.RoundTrip) {
+		switch rt.Verb() {
+		case kubeclient.VerbGet, kubeclient.VerbList, kubeclient.VerbWatch:
+			// reads are always allowed.
+			// note that while our pods/exec into the kube cert agent pod is a write request from the
+			// perspective of the Kube API, it is semantically a read request since no mutation occurs.
+			// we simply use it to fill a cache, and we need all pods to have a functioning cache.
+			// however, we do not need to handle it here because remotecommand.NewSPDYExecutor uses a
+			// kubeclient.Client.JSONConfig as input.  since our middleware logic is only injected into
+			// the generated clientset code, this JSONConfig simply ignores this middleware all together.
+			return
+		}
+
+		if isLeader.Load() { // only perform "expensive" test for writes
+			return // we are currently the leader, all actions are permitted
+		}
+
+		rt.MutateRequest(func(_ kubeclient.Object) error {
+			return ErrNotLeader // we are not the leader, fail the write request
```

Comment:
When this error is returned to controllers, they will start backing off -- if the other pod dies and they are elected leader, won't they be stuck in a very slow backoff state? Just wondering how quickly things will recover when a new leader is elected.

Reply:
I guess the controllers mostly still shouldn't be error-looping because all instances of the controller converge on the same desired state anyway? So this error shouldn't be spammed that much, only while things are coming online or after a spec change?

Reply:
I do not think this is an issue because all our controllers are written to limit writes to the Kube API (and even the ones that issue live reads are fine since that is always allowed).  Thus other than on initial start up / config changes, they will just perform a bunch of (mostly informer) reads on every sync.

To confirm this I checked the logs of pods that have been running this change for many hours and did find a minor bug in the `legacy-pod-cleaner-controller`.  Even with this bug the sync loop still ran every ~15 mins since that is the upper bound of the rate limiter.  I have pushed a fix.